x86_energy_sync_plugin.cpp

```diff
@@ -0,0 +1,654 @@
+         * This is necessary if the plugin gets reinitalisized and the
+         * synchronise funciton is not called, for example if PTF is used
+         */
+        this->responsible_thread = global_responsible_thread;
+        this->is_resposible = global_is_resposible_process;
+        logging::info() << "got local responsibility from global";
+
+        start_x86_energy();
+    }
+    
+    /**
+     * Destructor
+     *
+     * Stopping x86_energy
+     */
+    ~x86_energy_sync_plugin()
+    {
+        pid_t ptid = syscall(SYS_gettid);
+        if (this->is_resposible && (this->responsible_thread == ptid))
+        {
+            try
+            {
+                for(int i = 0; i < nr_packages; i++)
+                    source->fini_device(i);
+
+                /* pointer is obtained from rapl and can't be freed */
+                source = nullptr;
+            }
+            catch (std::exception& e)
+            {
+                logging::error() <<"Error occoured: " << e.what();
+            }
+
+            if (invalid_result_count > 0)
+            {
+                logging::warn() << "got " << invalid_result_count 
+                    << " invalid results at host: "<< this->hostname;
+                logging::warn() << "and " << valid_result_count 
+                    << " valid results at host: "<< this->hostname;
+            }
+        }
+        logging::debug()<< "plugin sucessfull finalized";
+
+    }
+    
+    /**
+     * You don't have to do anything in this method, but it tells the plugin 
+     * that this metric will indeed be used
+     */
+    void add_metric(x86_energy_sync_metric& m)
+    {
+        logging::info() << "adding x86_energy metric " << m.name();
+        logging::debug() << "adding counter for ptid:" << syscall(SYS_gettid);
+    }
+
+    /** Will be called for every event in by the measurement environment.
+     * You may or may not give it a value here.
+     *
+     * @param m contains the sored metric informations
+     * @param proxy get and save the results
+     *
+     * NOTE: In this implemenation we use a few assumptions:
+     * * scorep calls at every event all metrics
+     * * this metrics are called everytime in the same order
+     **/
+    template<typename P>
+    void get_current_value(x86_energy_sync_metric& m, P& proxy)
+    {
+        //retrun 0 if not responsible. Simulate PER_HOST.
+        pid_t ptid = syscall(SYS_gettid);
+        if(!this->is_resposible || (this->responsible_thread != ptid))
+        {
+            proxy.store((int64_t)0);
+            return;
+        }
+
+        //only possible error is that x86_energy was not initialized before
+        if (nr_packages < 0)
+        {
+            logging::debug() << "start_x86_energy was not called before";
+            invalid_result_count++;
+
+            /* no results can be obtained */
+            proxy.store((int64_t)0);
+            return;
+        }
+
+        if (m.quantity() != "E")
```

Comment:
Power is not implemented in this plugin version and also other quantities would be possible